### PR TITLE
feat: Bun Adapter

### PR DESCRIPTION
**Pull request by EamonHeffernan**
# Overview

# What is it?

- [x] Feature / enhancement
- [ ] Bug
- [x] Docs / tests / types / typos

# Description

Adds an adapter for use with Bun as it is being released in v1.0 today.
Bun's package manager cannot run Vite at the moment, therefore npm must be used to install, but bun can be used to run the dev and production servers.

# Use cases and why

- 1. One use case
- 2. Another use case

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/BuilderIO/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by birkskyum** [comment]
> Bun's package manager cannot run Vite at the moment

What specifically is the issue here?

**Comment by EamonHeffernan** [comment]
> > Bun's package manager cannot run Vite at the moment
> 
> What specifically is the issue here? Is there an open ticket in the Bun repo?

That might have been unclear sorry, if the packages are installed via `bun install` running `bun run dev` fails due to this issue with sharp https://github.com/lovell/sharp/issues/3511. If they are installed via `npm install`, `bun run dev` works fine.

File: starters/adapters/bun/src/entry.bun.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * Learn more about the Bun integration here:
+ * - https://qwik.builder.io/docs/deployments/bun/
+ * - https://bun.sh/docs/api/http
+ *
+ */
+import { createQwikCity } from "@builder.io/qwik-city/middleware/bun";
+import qwikCityPlan from "@qwik-city-plan";
+import { manifest } from "@qwik-client-manifest";
+import render from "./entry.ssr";
+
```
Comment (Mouvedia) [inline starters/adapters/bun/src/entry.bun.ts:15]: Is
```ts
declare global {
  interface QwikCityPlatform extends PlatformBun {}
}
```
necessary?

Comment (EamonHeffernan) [reply starters/adapters/bun/src/entry.bun.ts:15]: It's not in the deno entry file. It may be necessary though, I can add it if needed.

Comment (Mouvedia) [reply starters/adapters/bun/src/entry.bun.ts:15]: It seems to be just a matter of adding
```ts
/**
 * @public
 */
export declare interface PlatformBun {
    ssr?: true;
    // or
    // ssr: boolean;
}
```
in `index.d.ts`.

**Comment by mhevery** [comment]
Thank you for your contribution. Let us know when you think it is ready or if you want to review a specific area.

**Comment by EamonHeffernan** [comment]
> Thank you for your contribution. Let us know when you think it is ready or if you want to review a specific area.

Should be good to go. Just not sure on whether I needed to add those type defs, it worked for me including cli.
It's very basic, doesn't use any of bun's special build features which I assume would take work on vites part.

**Review by zanettin — COMMENTED** [review]
Hi @EamonHeffernan 
Just went over the code and left a short question. Thanks a lot for this addition 🙏

File: packages/docs/src/routes/docs/deployments/bun/index.mdx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+title: Bun Middleware | Deployments
+contributors:
+  - EamonHeffernan
+---
+
+# Bun Middleware
+
+Qwik City Bun middleware allows you to hook up Qwik City to a Bun server which uses the Bun Http API.
+
+## Installation
```
[the remark this answers is not in the document]
Comment (EamonHeffernan) [reply packages/docs/src/routes/docs/deployments/bun/index.mdx:11]: Hi, I've added much more detailed information about both the installation and the current issue with using bun install.

**Comment by zanettin** [comment]
thank u very much for the changes on the docs @EamonHeffernan 🙏 
wanted to test run the changes but maybe my setup was not accurate. can you quickly tell me how you've tested it?
in the best case we'd be able to test from `bun qwik add bun` on the whole flow 👍  thx again

**Comment by EamonHeffernan** [comment]
> thank u very much for the changes on the docs @EamonHeffernan 🙏 wanted to test run the changes but maybe my setup was not accurate. can you quickly tell me how you've tested it? in the best case we'd be able to test from `bun qwik add bun` on the whole flow 👍 thx again

My flow for getting it working was build with the docker environment, npm install on a new qwik project, copy dist for qwik and qwik-city to replace the existing folders for qwik and qwik city, then run bun run qwik add bun. Happy to work out what's going on if you can't recreate it.

**Comment by gioboa** [comment]
Close #4239

**Comment by zanettin** [comment]
Hi @EamonHeffernan 👋 
Thanks again for your help 🙏 looks like (at least me) have issues with `pnpm` and the described flow. I am able to follow these steps when using npm. i've attached the errors below. is this reproduceable on your side?
Bildschirmfoto 2023-09-14 um 22 28 09


and a tiny cosmetic thing: can we adjust the formatting of the CLI output? the box does not fit well atm and shouldn't the listed cmd `bun build.server` be `bun run build.server` in `New scripts added` instead? 👼 
Bildschirmfoto 2023-09-14 um 22 15 57

Thanks once again for this fantastic work 🙏 really really appreciated!

**Comment by EamonHeffernan** [comment]
> Hi @EamonHeffernan 👋 Thanks again for your help 🙏 looks like (at least me) have issues with `pnpm` and the described flow. I am able to follow these steps when using npm. i've attached the errors below. is this reproduceable on your side? Bildschirmfoto 2023-09-14 um 22 28 09
> 
> and a tiny cosmetic thing: can we adjust the formatting of the CLI output? the box does not fit well atm and shouldn't the listed cmd `bun build.server` be `bun run build.server` in `New scripts added` instead? 👼 Bildschirmfoto 2023-09-14 um 22 15 57
> 
> Thanks once again for this fantastic work 🙏 really really appreciated!

Hmm, I haven't seen that issue. My guess on why pnpm doesn't work is that copying the files over might not work as well with its symlinking system, although I might be wrong.  My hope would be that this would be resolved upon upload as it looks like when the folder contents are replaced it doesn't properly detect the other packages.

As for the formatting on the CLI, I'm not really sure how to change that, my changes shouldn't of messed with that and I believe it will have been an existing issue.

**Comment by zanettin** [comment]
thanks again for your feedback 🙏 let me do some final tests and discuss it with the core team.

**Comment by EamonHeffernan** [comment]
Related PR is #5165 that automatically adds a trusted dependency on `npm/pnpm/yarn/bun create qwik` so that bun as a package manager will work for qwik, regardless of the adapter being used.

**Comment by zanettin** [comment]
@EamonHeffernan i've fixed the CLI inputs within your PR. so imo we are good to go with this 🚀 
thanks again for your work on this an patience 🙏 

Bildschirmfoto 2023-09-16 um 21 26 12

**Review by zanettin — APPROVED** [review]
thx again 🙏 🎉

**Comment by EamonHeffernan** [comment]
> thx again 🙏 🎉

I've just realised there was something missing due to the changes that went through in the other pr. I've slightly changed the documentation to reflect this.

**Review by zanettin — APPROVED** [review]
thx for these adjustments as well 🙏

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
Thank you for your hard work and this excellent contribution. ❤️

**Comment by chuikoffru** [comment]
Hello, @EamonHeffernan

Thanks a lot for this adapter!

I faced with trouble of inner pages in Bun 1.0.4.

Step by step:
1) `bun create qwik@latest`, select Demo Site with demo pages
2) `bun run qwik add bun`
3) `bun run build`
4) `bun run serve`
5) Open localhost:3000, and main page working well
6) Go to http://localhost:3000/demo/flower - it's not ok. It's not working.

Could you fix please. Thanks again.

**Comment by EamonHeffernan** [comment]
> Hello, @EamonHeffernan
> 
> Thanks a lot for this adapter!
> 
> I faced with trouble of inner pages in Bun 1.0.4.
> 
> Step by step:
> 
>     1. `bun create qwik@latest`, select Demo Site with demo pages
> 
>     2. `bun run qwik add bun`
> 
>     3. `bun run build`
> 
>     4. `bun run serve`
> 
>     5. Open localhost:3000, and main page working well
> 
>     6. Go to http://localhost:3000/demo/flower - it's not ok. It's not working.
> 
> 
> Could you fix please. Thanks again.

Hi,
Very sorry that that got overlooked, I've tracked down the issues with that.
I've opened the pull request #5272 that resolves this issue.

**Pull request merged**